File: app/src/main/java/com/kent/university/privelt/base/BaseActivity.java
package com.kent.university.privelt.base;

import android.os.Bundle;
import android.view.MenuItem;

import com.kent.university.privelt.PriVELTApplication;
import com.kent.university.privelt.api.PasswordManager;
import com.kent.university.privelt.api.ServiceHelper;
import com.kent.university.privelt.di.DaggerPriVELTComponent;
import com.kent.university.privelt.di.PriVELTComponent;
import com.kent.university.privelt.di.RoomModule;
import com.kent.university.privelt.viewmodel.PriVELTViewModelFactory;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import butterknife.ButterKnife;

public abstract class BaseActivity extends AppCompatActivity {

    protected abstract int getActivityLayout();

    protected abstract void configureViewModel();

    protected abstract void configureDesign(@Nullable Bundle savedInstanceState);

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Associates the layout file to this class
        this.setContentView(this.getActivityLayout());

        // Using the ButterKnife library
        ButterKnife.bind(this);

        configureViewModel();

        configureDesign(savedInstanceState);
    }

    protected <T extends ViewModel> T getViewModel(final Class<T> className) {
        // Component
        final PriVELTComponent component = DaggerPriVELTComponent.builder().roomModule(new RoomModule(this)).build();

        // ViewModelFactory
        final PriVELTViewModelFactory factory = component.getViewModelFactory();

       return new ViewModelProvider(this, factory).get(className);
    }

    protected PasswordManager getIdentityManager() {
        return ((PriVELTApplication) getApplication()).getIdentityManager();
    }

    protected ServiceHelper getServiceHelper() {
        return ((PriVELTApplication)getApplicationContext()).getServiceHelper();
    }

    @Override
    protected void onResume() {
        super.onResume();
        ((PriVELTApplication) getApplicationContext()).setCurrentActivity(this);
    }


    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/kent/university/privelt/base/BaseActivity.kt
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.kent.university.privelt.base

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kent.university.privelt.PriVELTApplication
import com.kent.university.privelt.api.PasswordManager
import com.kent.university.privelt.di.DaggerPriVELTComponent
import com.kent.university.privelt.di.RoomModule

abstract class BaseActivity : AppCompatActivity() {
    protected abstract val activityLayout: Int
    protected abstract fun configureViewModel()
    protected abstract fun configureDesign(savedInstanceState: Bundle?)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Associates the layout file to this class
        this.setContentView(activityLayout)

        configureViewModel()
        configureDesign(savedInstanceState)
    }

    protected fun <T : ViewModel?> getViewModel(className: Class<T>): T {
        // Component
        val component = DaggerPriVELTComponent.builder().roomModule(RoomModule(this)).build()

        // ViewModelFactory
        val factory = component.viewModelFactory
        return ViewModelProvider(this, factory!!).get(className)
    }

    protected val identityManager: PasswordManager?
        get() = (application as PriVELTApplication).identityManager

    override fun onResume() {
        super.onResume()
        (applicationContext as PriVELTApplication).currentActivity = this
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}